File: python/dsa/data-structures/design/design-hash-map.py
class MyHashMap:
    def __init__(self): # @viz:init
        self.size = 8
        self.buckets = [[] for _ in range(self.size)]

    def _hash(self, key):
        return key % self.size

    def put(self, key: int, value: int) -> None: # @viz:put_start
        idx = self._hash(key) # @viz:hash
        bucket = self.buckets[idx] # @viz:put_bucket
        for i, (k, v) in enumerate(bucket): # @viz:check_exists
            if k == key: # @viz:check_key
                bucket[i][1] = value # @viz:update
                return # @viz:put_return
        bucket.append([key, value]) # @viz:insert

    def get(self, key: int) -> int: # @viz:get_start
        idx = self._hash(key) # @viz:get_hash
        bucket = self.buckets[idx] # @viz:get_bucket
        for k, v in bucket: # @viz:get_search
            if k == key: # @viz:get_check_key
                return v # @viz:get_found
        return -1 # @viz:get_not_found

    def remove(self, key: int) -> None: # @viz:remove_start
        idx = self._hash(key) # @viz:remove_hash
        bucket = self.buckets[idx] # @viz:remove_bucket
        for i, (k, v) in enumerate(bucket): # @viz:remove_search
            if k == key: # @viz:remove_check_key
                bucket.pop(i) # @viz:remove_done
                return # @viz:remove_return
        # Not found # @viz:remove_not_found

# Simulation end # @viz:finish

File: javascript/dsa/data-structures/design/design-hash-map.js
class MyHashMap {
  constructor() { // @viz:init
    this.size = 8;
    this.buckets = Array.from({ length: this.size }, () => []);
  }

  hash(key) {
    return key % this.size;
  }

  put(key, value) { // @viz:put_start
    const idx = this.hash(key); // @viz:hash
    const bucket = this.buckets[idx]; // @viz:put_bucket
    for (let i = 0; i < bucket.length; i++) { // @viz:check_exists
      if (bucket[i][0] === key) { // @viz:check_key
        bucket[i][1] = value; // @viz:update
        return; // @viz:put_return
      }
    }
    bucket.push([key, value]); // @viz:insert
  }

  get(key) { // @viz:get_start
    const idx = this.hash(key); // @viz:get_hash
    const bucket = this.buckets[idx]; // @viz:get_bucket
    for (let i = 0; i < bucket.length; i++) { // @viz:get_search
      if (bucket[i][0] === key) { // @viz:get_check_key
        return bucket[i][1]; // @viz:get_found
      }
    }
    return -1; // @viz:get_not_found
  }

  remove(key) { // @viz:remove_start
    const idx = this.hash(key); // @viz:remove_hash
    const bucket = this.buckets[idx]; // @viz:remove_bucket
    for (let i = 0; i < bucket.length; i++) { // @viz:remove_search
      if (bucket[i][0] === key) { // @viz:remove_check_key
        bucket.splice(i, 1); // @viz:remove_done
        return; // @viz:remove_return
      }
    }
    // Not found // @viz:remove_not_found
  }
} // @viz:finish